src/common/components/dateTimePicker/DateTimePicker.tsx

```diff
@@ -0,0 +1,303 @@
+import { useState, useRef, useEffect, Ref, useCallback } from "react";
+import {
+  IconCalendar,
+  IconAngleDown,
+  DateInput,
+  TimeInput,
+  Button,
+  IconCheck,
+  IconCross,
+} from "hds-react";
+import classNames from "classnames";
+import { useTranslation } from "react-i18next";
+import { format, isAfter, isBefore } from "date-fns";
+
+import useIntermediaryState from "../../hooks/useIntermediaryState";
+import { Locale } from "../../../config";
+import getIsDateValid from "../../utils/getIsValidDate";
+import { formatIntoDateTime, formatIntoDate } from "../../utils/time/format";
+import styles from "./dateTimePicker.module.scss";
+
+function getDate(value?: Date): string {
+  if (!value) {
+    return null;
+  }
+
+  return format(value, "d.M.y");
+}
+
+function getTime(value?: Date): string {
+  if (!value) {
+    return null;
+  }
+
+  return format(value, "H.m");
+}
+
+function getDateFromDateString(dateString?: string): Date | undefined {
+  if (!dateString) {
+    return;
+  }
+
+  const [D, M, Y] = dateString.split(".");
+
+  if (!D || !M || !Y) {
+    return;
+  }
+
+  return new Date(Number(Y), Number(M) - 1, Number(D));
+}
+
+function getDateFromDateAndTimeString(
+  dateString?: string,
+  timeString?: string
+): Date | undefined {
+  if (!dateString || !timeString) {
+    return;
+  }
+
+  const [D, M, Y] = dateString.split(".");
+  const [H, m] = timeString.split(":");
+  const hasAllDateParts = [D, M, Y, H, m].reduce(
+    (acc, part) => acc && Boolean(part),
+    true
+  );
+
+  if (!hasAllDateParts) {
+    return;
+  }
+
+  return new Date(Number(Y), Number(M) - 1, Number(D), Number(H), Number(m));
+}
+
+function validateDate(
+  date: Date | undefined | null,
+  dateToCompare: Date | undefined | null,
+  fnc: (date: Date, dateToCompare: Date) => boolean,
+  message: string
+): string {
+  if (!date || !dateToCompare) {
+    return "";
+  }
+
+  const isError = fnc(date, dateToCompare);
+
+  if (isError) {
+    return message;
+  }
+
+  return "";
+}
+
+function useOnOutsideClick<T extends HTMLElement>(
+  callback: () => void
+): Ref<T | null> {
+  const ref = useRef<null | T>(null);
+
+  useEffect(() => {
```

Comment:
I think that React docs recommend using `useEffect` when possible, and only using the layout version when it doesn't work:

https://reactjs.org/docs/hooks-reference.html#uselayouteffect

Here's more info from Kent Dodds: https://kentcdodds.com/blog/useeffect-vs-uselayouteffect/

Seems like it would be necessary when you inspect DOM content, for instance element dimensions.

Reply:
I remember now that I have had the problem with react testing library. Those event handlers ended up being out of sync and useLayoutEffect fixed that. Might never be issue in real world.

Reply:
The picker is tested with react-testing-library's `user-events` plugin. That should provide better event simulation. Not sure if that could make the difference here.